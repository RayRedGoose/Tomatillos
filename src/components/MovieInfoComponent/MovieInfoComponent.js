import React, { Component } from 'react';
import { connect } from 'react-redux';
import './MovieInfoComponent.scss'
import { Redirect } from 'react-router-dom';

class MovieInfoComponent extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isRedirected: false
    }
  }
  redirect = () => {
    this.setState({isRedirected: true})
  }

  render() {
    const { title, release_date, poster_path, overview, average_rating} = this.props.movie;
    const year = release_date.split('-')[0];

    const cardStyles = {
      backgroundImage: `url(${poster_path})`,
      backgroundSize: '100%',
      backgroundPosition: 'center',
      width: '230px',
      height: '300px'
    }

    return (
      (this.state.isRedirected)
        ? <Redirect to='/' />
        : <main className="single-movie-info">
          <button onClick={this.redirect}>&#8592; back to all movies</button>
          <div className='title-div'><h2>{`${title} (${year})`}</h2></div>
          <div className='img-container' style={cardStyles}></div>
          <section className='overview-box'>
            <h3>Avr.Rating: {average_rating.toFixed(1)}</h3>
            {this.props.login && <h3>Your rating: {average_rating}</h3>}
            <p>{overview}</p>
          </section>
        </main>
    )
  }
}

export const mapStateToProps = (state, props) => ({
  movie: state.movies.find(movie => movie.id === parseInt(props.id)),
  login: state.login
});

export default connect(mapStateToProps)(MovieInfoComponent);


// id: 3
// title: "Frozen II"
// poster_path: "https://image.tmdb.org/t/p/original//pjeMs3yqRmFL3giJy4PMXWZTTPa.jpg"
// backdrop_path: "https://image.tmdb.org/t/p/original//xJWPZIYOEFIjZpBL7SVBGnzRYXp.jpg"
// release_date: "2019-11-20"
// overview: "Elsa, Anna, Kristoff and Olaf head far into the forest to learn the truth about an ancient mystery of their kingdom."
// average_rating: 8
