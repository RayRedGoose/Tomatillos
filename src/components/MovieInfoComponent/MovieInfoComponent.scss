@import "../../variables.scss";
.single-movie-info {
  h2 {
    margin-top: 100px;
    margin-left: 350px;
    display: inline-block;
    padding: 20px;
    border: $main-accent-color solid 3px;
    background: $main-background-color-transparent;
    color: #fff;
    text-shadow: 0 0 24px #000;
  }
  &>button {
    margin: 30px 60px;
    position: absolute;
  }
  button {
    color: #fff;
    cursor: pointer;
    font-size: 18px;
    background: none;
    border: none;
    &:hover {
      transform: scale(1.1);
    }
  }
}

.img-container {
  position: absolute;
  left: 100px;
  top: 160px;
  z-index: 1;
  box-shadow: 0 0 48px 0 #000;
}

.overview-box {
  border: $main-accent-color solid 3px;
  background: $main-background-color-transparent;
  backdrop-filter: blur(4px);
  width: 55%;
  margin: auto;
  padding: 20px 20px 20px 90px;
  h3 {
    margin: 0;
    &:first-child {
      margin-bottom: 5px;
    }
  }
  &>p {
    line-height: 1.5em;
    min-height: 100px;
    max-height: 300px;
    overflow-x: scroll;
  }
  h3, p {
    color: $main-text-color;
  }
  footer {
    text-align: center;
    width: 100%;
    margin-top: 80px;
    button {
      margin: auto;
      background-color: $main-accent-color;
      border-radius: 20px;
      padding: 0.4em 2em;
    }
  }
}
