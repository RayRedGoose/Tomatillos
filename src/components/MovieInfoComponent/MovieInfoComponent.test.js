import React from 'react';
import { MovieInfoComponent, mapStateToProps } from './MovieInfoComponent';
import { shallow } from 'enzyme';

describe("MovieInfoComponent", () => {
  const mockMovieData = {
    id: 1,
    title: "Jumanji: The Next Level",
    poster_path: 'https://some.url/poster-image',
    backdrop_path: 'https://some.url/bg-image',
    release_date: '2019-12-04',
    overview: 'In Jumanji: The Next Level, the gang is back but the game has changed.',
    average_rating: 4
  };

  const mockRatingData = {
    id: 1,
    user_id: 1,
    movie_id: 1,
    rating: 7,
    created_at: "2019-12-16",
    updated_at: "2019-12-16"
  };

  describe("MovieInfoComponent component", () => {
    let container;

    const mockProps = {
      movie: mockMovieData,
      rating: mockRatingData,
      user: true,
      id: 1
    };

    beforeEach(() => {
      container = shallow(
        <MovieInfoComponent {...mockProps} />
      )
    });

    it("should match snapshot with all data passed in correctly", () => {
      expect(container).toMatchSnapshot();
    });

    it("should match snapshot if redirected state is true", () => {
      container.setState({isRedirected: true});
      expect(container).toMatchSnapshot();
    });

    it("should match snapshot if there is no user", () => {
      const containerWithoutUser = shallow(
        <MovieInfoComponent {...mockProps} user={false} />
      )
      expect(containerWithoutUser).toMatchSnapshot();
    });

    it("should match snapshot if there is user but no rating", () => {
      const containerWithoutRating = shallow(
        <MovieInfoComponent {...mockProps} rating={null} />
      )
      expect(containerWithoutRating).toMatchSnapshot();
    });
  });

  describe("mapStateToProps", () => {
    it('should return an object with the movie, rating and user object', () => {
      // Setup
      const mockState = {
        movies: [mockMovieData],
        ratings: [mockRatingData],
        error: '',
        user: {name: 'Ray'},
        selectedMovie: 1
      };

      const expected = {
        movie: mockMovieData,
        rating: mockRatingData,
        user: {name: 'Ray'}
      };

      // Execution
      const mappedProps = mapStateToProps(mockState);

      // Expectation
      expect(mappedProps).toEqual(expected);
    });
  });
});
