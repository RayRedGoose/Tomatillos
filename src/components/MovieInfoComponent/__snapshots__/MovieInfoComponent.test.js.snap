// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieInfoComponent MovieInfoComponent component should match snapshot if redirected state is true 1`] = `
<Redirect
  to="/"
/>
`;

exports[`MovieInfoComponent MovieInfoComponent component should match snapshot if there is no user 1`] = `
<main
  className="single-movie-info"
>
  <button
    onClick={[Function]}
  >
    ← back to all movies
  </button>
  <div
    className="title-div"
  >
    <h2>
      Jumanji: The Next Level (2019)
    </h2>
  </div>
  <div
    className="img-container"
    style={
      Object {
        "backgroundImage": "url(https://some.url/poster-image)",
        "backgroundPosition": "center",
        "backgroundSize": "100%",
        "height": "300px",
        "width": "230px",
      }
    }
  />
  <section
    className="overview-box"
  >
    <h3>
      Avr.Rating: 
      4.0
    </h3>
    <p>
      In Jumanji: The Next Level, the gang is back but the game has changed.
    </p>
    <footer />
  </section>
</main>
`;

exports[`MovieInfoComponent MovieInfoComponent component should match snapshot if there is user but no rating 1`] = `
<main
  className="single-movie-info"
>
  <button
    onClick={[Function]}
  >
    ← back to all movies
  </button>
  <div
    className="title-div"
  >
    <h2>
      Jumanji: The Next Level (2019)
    </h2>
  </div>
  <div
    className="img-container"
    style={
      Object {
        "backgroundImage": "url(https://some.url/poster-image)",
        "backgroundPosition": "center",
        "backgroundSize": "100%",
        "height": "300px",
        "width": "230px",
      }
    }
  />
  <section
    className="overview-box"
  >
    <h3>
      Avr.Rating: 
      4.0
    </h3>
    <p>
      In Jumanji: The Next Level, the gang is back but the game has changed.
    </p>
    <footer>
      <Connect(RatingBox)
        movie_id={1}
      />
    </footer>
  </section>
</main>
`;

exports[`MovieInfoComponent MovieInfoComponent component should match snapshot with all data passed in correctly 1`] = `
<main
  className="single-movie-info"
>
  <button
    onClick={[Function]}
  >
    ← back to all movies
  </button>
  <div
    className="title-div"
  >
    <h2>
      Jumanji: The Next Level (2019)
    </h2>
  </div>
  <div
    className="img-container"
    style={
      Object {
        "backgroundImage": "url(https://some.url/poster-image)",
        "backgroundPosition": "center",
        "backgroundSize": "100%",
        "height": "300px",
        "width": "230px",
      }
    }
  />
  <section
    className="overview-box"
  >
    <h3>
      Avr.Rating: 
      4.0
    </h3>
    <h3>
      Your rating: 7
    </h3>
    <p>
      In Jumanji: The Next Level, the gang is back but the game has changed.
    </p>
    <footer />
  </section>
</main>
`;
